providers/asrockrack/helpers.go

```diff
@@ -0,0 +1,408 @@
+	Action   string `json:"action,omitempty"`
+	Progress string `json:"progress,omitempty"`
+	State    int    `json:"state,omitempty"`
+}
+
+// BIOS upgrade commands
+// 2 == configure
+// 3 == apply upgrade
+type biosUpdateAction struct {
+	Action int `json:"action"`
+}
+
+// 1 Set BMC to flash mode and prepare flash area
+// at this point all logged in sessions are terminated
+// and no logins are permitted
+func (a *ASRockRack) setFlashMode() error {
+
+	endpoint := "api/maintenance/flash"
+
+	_, statusCode, err := a.queryHTTPS(endpoint, "PUT", nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if statusCode != 200 {
+		return fmt.Errorf("Non 200 response: %d", statusCode)
+	}
+
+	a.resetRequired = true
+
+	return nil
+}
+
+// 2 Upload the firmware file
+func (a *ASRockRack) uploadFirmware(endpoint string, filePath string) error {
+
+	// setup a buffer for our multipart form
+	var form bytes.Buffer
+	w := multipart.NewWriter(&form)
+
+	// create form data from update image
+	fwWriter, err := w.CreateFormFile("fwimage", path.Base(filePath))
+	if err != nil {
+		return err
+	}
+
+	// open file handle
+	fh, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+
+	// copy file contents into form payload
+	fReader := bufio.NewReader(fh)
+	_, err = io.Copy(fwWriter, fReader)
+	if err != nil {
+		return err
+	}
+
+	// multi-part content type
+	headers := map[string]string{"Content-Type": w.FormDataContentType()}
+
+	// close multipart writer - adds the teminating boundary.
+	w.Close()
+
+	// POST payload
+	_, statusCode, err := a.queryHTTPS(endpoint, "POST", form.Bytes(), headers)
+	if err != nil {
+		return err
+	}
+
+	if statusCode != 200 {
+		return fmt.Errorf("Non 200 response: %d", statusCode)
+	}
+
+	return nil
+}
+
+// 3. Verify uploaded firmware file - to be invoked after uploadFirmware()
+func (a *ASRockRack) verifyUploadedFirmware() error {
+
+	endpoint := "api/maintenance/firmware/verification"
+
+	_, statusCode, err := a.queryHTTPS(endpoint, "GET", nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if statusCode != 200 {
+		return fmt.Errorf("Non 200 response: %d", statusCode)
+	}
+
+	return nil
+
+}
+
+// 4. Start firmware flashing process - to be invoked after verifyUploadedFirmware
+func (a *ASRockRack) upgradeBMC() error {
+
+	endpoint := "api/maintenance/firmware/upgrade"
+
+	// preserve all configuration during upgrade, full flash
+	pConfig := &preserveConfig{PreserveConfig: 1, FlashStatus: 1}
+	payload, err := json.Marshal(pConfig)
+	if err != nil {
+		return err
+	}
+
+	headers := map[string]string{"Content-Type": "application/json"}
+	_, statusCode, err := a.queryHTTPS(endpoint, "PUT", payload, headers)
+	if err != nil {
+		return err
+	}
+
+	if statusCode != 200 {
+		return fmt.Errorf("Non 200 response: %d", statusCode)
+	}
+
+	return nil
+
+}
+
+// 4. reset BMC
+func (a *ASRockRack) reset() error {
+
+	endpoint := "api/maintenance/reset"
+
+	_, statusCode, err := a.queryHTTPS(endpoint, "GET", nil, nil)
```

Comment:
I got a 404 from this in playing around with this code; looking at the doc, it looks like it's supposed to be a POST.  I tried it after a simple `s/GET/POST/` on this line and it seemed to work.

Reply:
thanks! fixed in b3b2d240